### ISSUE TITLE
Sailsjs(~0.10.5) + Socket.io (0.9.14) + PM2 deploy with multi instances issue

### ISSUE THREAD

**Issue by handleror** [balderdashy/sails#2538]
I have got a big issue when deploy sails code (0.10.5) with pm2 multi instances (>=2):
Socket.emit always send multi times and raise:
> [API-2 (err)] Error: Can't set headers after they are sent.
> [API-2 (err)]   at ServerResponse.OutgoingMessage.setHeader (http.js:689:11)

I did see the problem in these links:
[socket.io!](https://github.com/Automattic/socket.io/issues/1636)
[pm2!](https://github.com/Unitech/PM2/issues/637) 

But there is no way to upgrade socket.io in sails?

**Comment by sgress454** [comment]
Hi @handleror, try giving the master branch of Sails a try?  It uses socket.io v1.0.

**Comment by handleror** [comment]
Hi @sgress454, I tried using master code, but got another issue with 
> warn: A socket connected, but the session could not be loaded to pass to configured handler: `sails.config.sockets.onConnect()`.  Will run handler with a fake, empty session.  Details:
> Error: Session could not be loaded 
> at _createError (node_modules/sails/lib/hooks/session/index.js:271:21)
> at Object._onImmediate (node_modules/sails/lib/hooks/session/index.js:274:13)
> at processImmediate [as _immediateCallback] (timers.js:330:15)
> { [Error: Session could not be loaded] code: 'E_SESSION' }
What's alter method for onConnect? (I need handle socket after connected) 
Thanks

**Comment by sgress454** [comment]
`onConnect` is deprecated in the upcoming version.  What we suggest is to place the code you formerly had in `onConnect` in a controller action, and have your client-side code do a request to the server using `io.socket.get` after the socket connects.  It might seem wasteful to do another roundtrip but it's a lot more reliable and it keeps all your code in the expected places.

**Comment by handleror** [comment]
Thank @sgress454, but I got another issue in integrating socket.io + redis (socket, adapter memory works well):
1. Socket could not join any room, this is exception:
> node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/base.js:246
>        throw message;
>             ^
> RangeError: Maximum call stack size exceeded

\2. Sometimes, could not start server with error:
> error: Error: The hook `orm` is taking too long to load.
> Make sure it is triggering its `initialize()` callback, or else set `sails.config.orm._hookTimeout to a higher value (currently 10000)
> at [object Object].tooLong as _onTimeout (node_modules/sails/lib/app/private/loadHooks.js:92:21)
  at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)

This is config on socket.js:
>   adapter: 'socket.io-redis',
>  adapterOptions: {
>    port: 6379,
>    key: 'socket',
>    db: 2,
>    // socket: "tmp/redis.sock"
>  },

**Comment by kaasdude** [comment]
@sgress454 Are you saying the "onConnect" and "onDisconnect" in config/socket.js will disappear eventually? 

I think the event that a client disconnects is still a great feature, because one can do stuff with it.
When a client disconnects, its not always because the client wants to disconnect, maybe the internet connection was lost, or laptop battery drainages... :-D

Also its great to see if its an existing user, when one tries to reconnect with the socket, the session is pretty handy to check if we're connecting again, and just have the same user again.

**Comment by sgress454** [comment]
@handleror you might have an old version of the `sails-hook-sockets` dependency--I think the `maximum call stack` issue is fixed now (pinging @mikermcneil for confirmation).  As far as the hook timeout, we've introduced the ability to set these globally with `sails.config.hookTimeout` in the latest build, and bumped the default up to 20 seconds.

@kaasdude totally agree that the disconnect event is valuable.  `onDisconnect` is being deprecated, but only in favor of a slightly easier `afterDisconnect` method.  See [this section in the migration guide for v0.11](https://github.com/balderdashy/sails/blob/master/0.11-migration-guide.md#ondisconnect-lifecycle-callback) for details.

**Comment by CWyrtzen** [comment]
Did v0.11 clear up the problem?

**Comment by kaasdude** [comment]
Sorry I didn't reply earlier, we stick to .10.latest because we needed to stay at socket.io version 0.9. This because we developed our mobile apps to connect with socket.io based on 0.9...


So maybe later, in a feature/upgrade session ;)

The problem also is that I'm involved in other projects, but when the moment arrives, I certainly wil migrate to 0.11+  

There is also the thing I submitted a ticket for, the Socket.io <=> redis relationship. Once this problem gets solved, we will will consider upgrading, see : https://github.com/balderdashy/sails/issues/2593

Regards.

**Comment by sailsbot** [comment]
Thanks for posting, @handleror.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by majgis** [comment]
Sticky cluster claims to have 10x the perf of sticky session.  https://www.npmjs.com/package/sticky-cluster  I wish there was an option with PM2 to enable sticky sessions.  I need to support websockets on legacy browsers.  Does anyone know of another library similar to PM2 that has support for sticky sessions?
